Handle `NaN` values in FormulaEngine

daniel-zullo-frequenz opened the pull request:
While you are at this, it should probably be a good idea to also treat *NaN*'s specially (probably the same as `None`). You can check using `math.isnan()`.

(same for all steps)

**OR**, maybe replace `None` with *NaN* and it should have the same effect of propagating `None` manually as you are doing here, but it will be done automatically. Then you only need to convert the final *NaN* in the stack to `None` and you are set. This option is easier to write (as you don't have to touch every step) but it might be harder read (to figure out what's going on).

_Originally posted by @leandro-lucarella-frequenz in https://github.com/frequenz-floss/frequenz-sdk-python/pull/95#discussion_r1038228733_

daniel-zullo-frequenz commented:
@sahas-subramanian-frequenz what is the rationale to treat float `nan` as `None`. In any case if we try `nan` in a special way, should we do the same for `inf`?

daniel-zullo-frequenz commented:
I'd like to check with you if this is what you have in mind to handle None/NaN, please see the example below

```python
from math import isnan, isinf

class Adder(FormulaStep):
    """A formula step for adding two values."""

    def __repr__(self) -> str:
        """Return a string representation of the step.

        Returns:
            A string representation of the step.
        """
        return "+"

    def apply(self, eval_stack: List[Optional[float]]) -> None:
        """Extract two values from the stack, add them, push the result back in.

        Args:
            eval_stack: An evaluation stack, to apply the formula step on.
        """
        val2 = float("NaN") if (val := eval_stack.pop()) is None else val
        val1 = float("NaN") if (val := eval_stack.pop()) is None else val
        res = val1 + val2
        if isnan(res) or isinf(res):
            res = None
        eval_stack.append(res)
  ```

shsms commented:
Hey @daniel-zullo-frequenz ,  I think that makes sense,  but I wonder what kind of perf cost this would have.  Maybe we should write some benchmark scripts for the data pipeline first.  We can talk about that this afternoon.

But another option for implementing this,  is to just check incoming values in the `MetricFetcher` if they are `NaN` or `INF`, and replace those with `None` in that step alone.

shsms commented:
Or like Luca originally mentioned,  we could replace `None` values with `NaN`, or `0` in the `MetricFetcher`, and remove all the special `None` handling in the other steps,  and in the engine's `apply` function,  replace the output with `None` if it was `NaN`.  Maybe that's be most efficient approach.

daniel-zullo-frequenz commented:
That makes sense. I'll work out the changes and push a MR. Thanks for the clarification!

daniel-zullo-frequenz committed:
Handle NaN/INF values in FormulaEngine

The MetricFetcher propagates NaN to handle None values
when None values are not treat from the stream as 0s.
So that any FormulaStep can compute the results without
checking for None on each value involved in the FormulaStep.
However the final result is written as None rather than NaN/INF
in the FormulaEngine.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

shsms commented on src/frequenz/sdk/timeseries/logical_meter/_formula_steps.py:
```diff
@@ -57,10 +58,9 @@ def apply(self, eval_stack: List[Optional[float]]) -> None:
         val1 = eval_stack.pop()
-        if val1 is None or val2 is None:
+        res = val1 + val2
+        if (isnan(res) or isinf(res)) and (len(eval_stack) == 0):
             res = None
```
I think we don't have to do this check at the end of each step.  It can be moved to the top-level apply (in the `FormulaEngine` class, from which the `apply` of the individual steps are called),  where we can do it once, after all the steps have finished.

daniel-zullo-frequenz replied:
Right, I was not sure about this but it makes sense as FormulaEngine knows if all steps are done. I'll update it and call for a review after that

shsms commented on src/frequenz/sdk/timeseries/logical_meter/_formula_steps.py:
```diff
@@ -265,7 +262,18 @@ def apply(self, eval_stack: List[Optional[float]]) -> None:
-        if self._next_value.value is None and self._nones_are_zeros:
-            eval_stack.append(0.0)
+
+        if len(eval_stack) and eval_stack[-1] is None:
+            if self._nones_are_zeros:
+                eval_stack[-1] = 0.0
+            else:
+                eval_stack[-1] = float("NaN")
```
if individual steps don't insert `None`, then this check can go away.

daniel-zullo-frequenz replied:
indeed, this was only added as a mechanism to handle the case where FormulaStep sets None

christianparpart commented on src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py:
```diff
@@ -154,7 +155,11 @@ async def _apply(self) -> Sample:
             raise RuntimeError(f"Formula application failed: {self._name}")
 
-        return Sample(metric_ts, eval_stack[0])
+        res = eval_stack.pop()
```
In the old code, it used to be `eval_stack[0]`. I do not know too much about it, but to me that reads like `[0]` is only getting the front element but not popping it.
In this code, the value is not just read but also popped. I wonder why CI is not catching this change. Might be me looking at the wrong part though :)

daniel-zullo-frequenz replied:
`eval_stack` is a `List` defined at the local scope of the function, it will cease to exist once the function returns

shsms replied:
> but to me that reads like [0] is only getting the front element but not popping it.

also if you look at line 155,  we are ensuring that there's only a single value after all formula steps are applied.  So `[0]` and `.pop()` are pretty much the same, in this case.

christianparpart commented on src/frequenz/sdk/timeseries/logical_meter/_formula_steps.py:
```diff
@@ -57,10 +58,7 @@ def apply(self, eval_stack: List[Optional[float]]) -> None:
-            res = None
-        else:
-            res = val1 + val2
+        res = val1 + val2
```
How is this change related to the above change? I wonder because reading the above change, that looks like it's increasing the chance to get `None` now, but in this line the `None` checks are removed and a value is now always expected to operate on operator`+`/`-`/`*`/`/` (sharing the question also for the code path below :) ).

daniel-zullo-frequenz replied:
The idea/goal is the MetricFetcher only propagates NaN or 0s (based on the configuration) rather than `None` such us `FormulaStep` doesn't need to deal with `None`

christianparpart commented on src/frequenz/sdk/timeseries/logical_meter/_formula_steps.py:
```diff
@@ -265,7 +254,12 @@ def apply(self, eval_stack: List[Optional[float]]) -> None:
+            if self._nones_are_zeros:
+                eval_stack.append(0.0)
+            else:
+                eval_stack.append(float("NaN"))
```
NB: Python >= 3.10 introduces `math.nan` for symbolic access to `NaN`.

daniel-zullo-frequenz replied:
Then it can/should be updated once 3.8 and 3.9 are no longer supported

sahas-subramanian-frequenz reviewed (dismissed):
LGTM!

daniel-zullo-frequenz commented:
@sahas-subramanian-frequenz I'm afraid I've lost your approval when re-signing the commit to be verified. Can you please re-approve it?

daniel-zullo-frequenz commented:
We've recently found out there is an issue with the `Averager` formula, I'll fix it and add a separate a test case with all inputs having values,  and one test with fewer with values.

daniel-zullo-frequenz committed:
Add tests for FormulaStep Averager

The tests for FormulaStep Averager were missing.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
It turned out `MetricFetcher.apply()` is not called/reached out when the FormulaStep `Averager`  is performed, so the current patch/code was actually correct. However I have added unit tests for `Averager` in a separate commit.
